File: Recursion and Backtracking/cpp/printSubsequences.cpp
// Problem Statement: Given a string, find all the possible subsequences of the string.

// Problem Link: https://www.codingninjas.com/studio/problems/print-subsequences_8416366?utm_source=striver&utm_medium=website&utm_campaign=a_zcoursetuf

#include <iostream>
#include <vector>
#include <string>
using namespace std;

vector<string> generateSubsequences(string s) {
    if (s.length() == 0) {
        vector<string> bs;
        bs.push_back("");
        return bs;
    }

    char ch = s[0];
    string ss = s.substr(1);

    vector<string> sa = generateSubsequences(ss);
    vector<string> res;

    for (string str : sa) {
        res.push_back(str);
        res.push_back(ch + str);
    }

    return res;
}

File: Recursion and Backtracking/python/printSubsequences.py
# Problem Statement: Given a string, find all the possible subsequences of the string.

# Problem Link: https://www.codingninjas.com/studio/problems/print-subsequences_8416366?utm_source=striver&utm_medium=website&utm_campaign=a_zcoursetuf

def generate_subsequences(s):
    if len(s) == 0:
        return [""]
        
    ch = s[0]
    ss = s[1:]
    
    sa = generate_subsequences(ss)
    res = []
    
    for string in sa:
        res.append(string)
        res.append(ch + string)
        
    return res
